### PR TITLE
New TCS function to get speed post information (including warning speed posts)

### PR DESCRIPTION
**Pull request by Sharpe49**
[Trello Card](https://trello.com/c/S38d2RoV/482-further-tcs-handles-related-to-signals-and-sound-events)
Initially part of PR 293 but now split into its own PR.

### CODE REVIEW

**Comment by Csantucci** [comment]
This problem http://www.elvastower.com/forums/index.php?/topic/35165-problems-with-speed-restriction-since-u20210511-0251/ has arisen when this PR has become part of the unstable release.

**Comment by Sharpe49** [comment]
This commit should improve the situation.

**Comment by Sharpe49** [comment]
The problem has been fixed (http://www.elvastower.com/forums/index.php?/topic/35165-problems-with-speed-restriction-since-u20210511-0251/page__view__findpost__p__272364)

File: Source/Orts.Simulation/Common/Scripting/TrainControlSystem.cs
Comment (YoRyan) [inline Source/Orts.Simulation/Common/Scripting/TrainControlSystem.cs:678]: I suggest using `SpeedPostShapeName` as the property name here. It would be more descriptive.

File: Source/Orts.Simulation/Simulation/Signalling/Signals.cs
Comment (YoRyan) [inline Source/Orts.Simulation/Simulation/Signalling/Signals.cs:1278]: Indexing a list like this is a linear O(n) operation. A dictionary could accomplish the same thing in constant-time.

I realize you were copying the existing signal ref code, so I'm not going to press the issue. It might be something to revisit later.

File: Source/Orts.Simulation/Common/Scripting/TrainControlSystem.cs
Comment (Sharpe49) [reply Source/Orts.Simulation/Common/Scripting/TrainControlSystem.cs:678]: If, one day, we add the possibility to add new signals or speed posts from packages (which seems to be something that the virtual filesystem concept can manage to do), I think we would have config files (in JSON for example) for each new signal or each new speed post.
In that context, I think the term "speed post type name" is more generic and can be reused later for speed posts with a JSON config file.

File: Source/Orts.Simulation/Simulation/Signalling/Signals.cs
Comment (Sharpe49) [reply Source/Orts.Simulation/Simulation/Signalling/Signals.cs:1278]: I am not really good at optimizing code. If you have any proposals, I will gladly take them into account.

**Comment by Sharpe49** [comment]
Ok, I agree with your modifications.

File: Source/Orts.Simulation/Common/Scripting/TrainControlSystem.cs
Comment (YoRyan) [reply Source/Orts.Simulation/Common/Scripting/TrainControlSystem.cs:678]: What gets me is that you have an equivalent SignalFeatures property, MainHeadSignalTypeName, that communicates the MSTS signal type: "NORMAL", "DISTANCE", "SPEED", etc. To a user who doesn't understand your intentions for SpeedPostTypeName, it looks like a completely different piece of data that, for some reason, shares the same name.

To me, the speed post **type name** would be something more like the **TrainObjectType.SpeedItemType** enum, which can be one of "Standard", "TempRestrictedStart", and "TempRestrictedResume". If you want access to the shape name, then I still think it would be better to use an unambiguous "SpeedPostShapeName" property. (Eventually, when we get new file formats and editor support (ugh), we can introduce custom identifiers.)

Comment (Sharpe49) [reply Source/Orts.Simulation/Common/Scripting/TrainControlSystem.cs:678]: > What gets me is that you have an equivalent SignalFeatures property, MainHeadSignalTypeName, that communicates the MSTS signal type: "NORMAL", "DISTANCE", "SPEED", etc. To a user who doesn't understand your intentions for SpeedPostTypeName, it looks like a completely different piece of data that, for some reason, shares the same name.

No, it contains the signal type name of the SIGCFG file, not the signal function type:
image

> To me, the speed post **type name** would be something more like the **TrainObjectType.SpeedItemType** enum, which can be one of "Standard", "TempRestrictedStart", and "TempRestrictedResume". If you want access to the shape name, then I still think it would be better to use an unambiguous "SpeedPostShapeName" property. (Eventually, when we get new file formats and editor support (ugh), we can introduce custom identifiers.)

That would be the function type of the speed post.

Comment (YoRyan) [reply Source/Orts.Simulation/Common/Scripting/TrainControlSystem.cs:678]: Huh! I totally misread that code.

Okay, then no objections from me. Thanks for clarifying this.

**Review by YoRyan — APPROVED** [review]
Approving! Do new TCS functions need ORMT approval before merging? @cjakeman

**Comment by Sharpe49** [comment]
> Approving! Do new TCS functions need ORMT approval before merging? @cjakeman

Any news about merging this?

**Comment by cjakeman** [comment]
Hi Ryan,

> Approving! Do new TCS functions need ORMT approval before merging? @cjakeman

That's not quite how it works (or should work). New features need a Trello card or a Blueprint and ORMT approves the new feature as set out in [the Contributing document](https://github.com/openrails/openrails/blob/master/Docs/Contributing.md). 
Contributors should be linking their PRs back to a Trello card or a blueprint so it is easy for a reviewer, such as yourself, to check that the feature has ORMT approval.

Hi Cedric,

I don't see a link here, but I find a rather dated [Trello card](https://trello.com/c/FlBnKn2h/446-expand-potential-of-tcs-scripts) and a similar [Blueprint](https://blueprints.launchpad.net/or/+spec/expand-potential-of-tcs-scripts). 

I think it's better if we follow the procedures as far as possible, so please would you write a more focussed card or Blueprint and link the PR to it. Then I will be happy to approve it and merge the PR.

I did notice that no update to the Manual was included. Is that not needed for this function?

Best wishes,

**Comment by Sharpe49** [comment]
Hi @cjakeman,

I agree, but one of the problem is that we don't see a lot of updates of the Trello cards and blueprints. So, we tend to think that those are a bit left abandonned. Many functionnalities have been integrated to the master branch and most of them are in a pending approval state on Launchpad.
If we respect as much as possible the procedures, the development will be slowed down a lot. For example, one of my blueprints of 2020 is still not approved while the code is already in the master branch. The ORMT must be more reactive to what's happening on the different platforms for the procedures to be used properly.
If we (developers and ORMT) can't respect those procedures, maybe they have to change (for example, delegating blueprints validation and cards update for some subsystems to people that work a lot on these subsystems).

About this pull request, this is more linked to the pull request #293 and in this message (https://github.com/openrails/openrails/pull/293#issuecomment-736365911), we moved the speed post information subject to a later commit.

About the manual, this function works similarly as the NextGenericSignalFeatures function. And this function was not described in the manual.

**Comment by cjakeman** [comment]
Hi Cedric,

Yes, I can see that a backlog of Trello cards and Blueprints is not encouraging and slows down the work. We'll try to be more responsive and, now that Open Rails has gathered so much functionality, we might consider delegating sub-systems. Another option is to enlarge ORMT.

I've approved [your Blueprint from 2020](https://blueprints.launchpad.net/or/+spec/power-supply-enhancements).

Perhaps documenting the TCS functions deserves its own project.

I've added the link for this PR to #293 and to the Trello card and merged this PR.

Thanks and best wishes,

**Pull request merged**
